### DISCUSSION TITLE
Add "Favorites" to connect dialog

### DISCUSSION THREAD

**Discussion by dcorson-ticino-com** [jamulussoftware/jamulus#1511] [UI Related]
My first attempt to add favorites by making the connect dialog tabbed and having a similar display to the existing server list for the favorites turned out to be a very big code change.  I am now thinking through how to do a "minimally invasive" favorites feature.

Present thoughts are for the following:

1. in the connect dialog rename the "Server Address" combobox to "Favorites"
2. add a button to the dialog "Add to Favorites"
3.  the combobox retains the possibility to add entries manually, clicking the Add to Favorites button puts the entry selected in the server list in the combobox list.
4. maybe it would be useful to bring the last used entry of the combobox to the top of the list each time.
5. as the favorites list has a determined length the oldest unused entry eventually falls off the end of the list.  Entries can also be removed with a mouse right click.

What do you think?  Does that fulfill the needs for Favorites?  What could be improved?

**Comment by gilgongo** [comment]
(I took the liberty of numbering your points so they can be referenced, sorry)

Assuming the dialogue you refer to in 2 is the connect window, then would someone take the decision to make a server a favourite _before_ they joined it? I would have thought it was more logical to make that decision after they had joined and read any welcome message or licence notice. In which case, that implies a "favourite" function would be available from the main window, perhaps? 

4 seems sensible. 

BTW I think the context menu for removing is the wrong decision, since it's literally the only Jamulus function we have on right click. So if somebody finds it, they will be forever right-clicking around the app wondering if there are other hidden functions. Best practice with context menus is to have **all** (or a constant subset) of functions available, or none. Better in this case to have simply had a delete icon (with explanatory tool tip) next to the current entry in the input field I think.

**Reply by gene96817** [reply]
How do you handle favorites of different genre? Especially, how do you update ping times with multiple genres?

Regarding 5, how soon will you declare an entry unused? What if I have a server I only visit once a month or a few times a year (e.g. 90 days)?

**Reply by gilgongo** [reply]
> how soon will you declare an entry unused?

Assuming a fixed length to the list (10? 20?) the entry with the oldest last selected date just falls off the bottom I think.

**Reply by dcorson-ticino-com** [reply]
I am not trying to recreate the normal server list with ping times and user nrs. and max users.
I did try once and decided it is at least highly impractical if not impossible with the present protocol and separation of data.
I am extending the functionality of the Server Address combobox by adding a button that adds the info of the presently selected server to the list in the combobox.  As such I am making a detour around the genres.  A hidden addition to the combobox will make it possible to use the server name to select although it is not an URL as has been required up 'til now.

As to 5) I will use the same method as in the PR to be included in the next version adding one line to the already existing context menu for the combobox for delete server.  
Otherwise when an entry is used it is moved to the top of the list as is any new entry.  The list is 12 members long (at the moment) any more than that falls off the end of the list, as gilgongo correctly observed.  The most used servers will always stay in the list, but those that are seldom used are in danger of falling off.

@gene96817 What do you think of using a selected server in the connect menu to add to the list as compared to using the server one is currently connected to with a button on the main window?  I originally put a button on the main window, but thought it maybe needed more explanation than a button in the connect dialog.  I don't know.

@gilgongo Note that all the editable fields in Jamulus already have a right click context menu.  That is nothing new.

**Reply by gene96817** [reply]
> What do you think of using a selected server in the connect menu to add to the list as compared to using the server one is currently connected to with a button on the main window?

I agree that selecting/marking a server to be a favorite from the connection list would be more obvious. Also, this eliminates the need to make a connection first to set a favorite.

Currently the Server Address combobox accepts IP addresses or host domain names. For the favorites list, will this be IP addresses. Mostly, I don't memorize the IP addresses. It wouldn't be very nice to memorize 10 of my favorite server's IP address. The server name is much more user friendly. Does this mean you plan to enable the Server Address combobox to accept either server names or IP addresses?

**Reply by gilgongo** [reply]
I meant it's the only **Jamulus** function we have on right click. We should have them consistently for all relevant contexts if we have them at all, because that's what context menus are for (so right-click on a fader, get mute, solo, group, help etc. Right-click on another item get help on its own, etc.). We don't really have enough contexts to warrant their use though, even if we do get them "for free" from the OS/Qt on fields (which deploys them consistently for fields, as you point out). Occasional "hidden" context commands are not good design when more usable alternatives are available.

**Reply by dcorson-ticino-com** [reply]
>Does this mean you plan to enable the Server Address combobox to accept either server names or IP addresses?

Yes, you can enter an IP or URL, but Add to Favorite places a server name.  It won't work if you type in a server name as that won't give the also needed IP (that is stored, but hidden).

Maybe this isn't such a good idea after all to overload the Server Address field with the Favorites.  A separate combobox may be a better solution.

**Reply by dcorson-ticino-com** [reply]
>so right-click on a fader, get mute, solo, group

I don't follow you.  Why would I right click on something to get a function that is already there ?  I right click to find functions that are not the every day obvious stuff that already has its own UI manifestation.
Most of the controls already have "What's this" on right click.  It's not like we never use a right click.

**Reply by gene96817** [reply]
> Maybe this isn't such a good idea after all to overload the Server Address field with the Favorites. 

I agree that the mixed mode could confuse a non-technical user.

**Reply by gilgongo** [reply]
> Why would I right click on something to get a function that is already there?

Exactly my point :-) Apps that have context menus (eg drawing tools) typically have large numbers of functions, some of which are applicable to specific contexts, and users use them all the time. We have very little of that, so we shouldn't be adopting a convention just for _one thing_ when a clear alternative is available just like all our other functions (buttons, links). If one Jamulus feature is "hidden" in a context menu, how does the user know whether there are others elsewhere?

This is what I meant about the delete on the drop-down BTW.


https://user-images.githubusercontent.com/4561747/114667771-ccb1ed80-9cf7-11eb-80d1-2405cb8446b4.mp4
